scripts/usb_reset.py

```diff
@@ -14,46 +14,266 @@
+    try:
+        with open(cg_dir + file_name, "w", 0) as f:
+            ctl = get_ctl(path, "rw")
+            if ctl:
+                f.write(ctl)
+    except IOError as e:
+        # log and continue
+        log.debug("Failed to {} {}: {}".format(
+            "allow" if allow else "deny", path, str(e)))
+
+
+def setup_cgroup(dom_id, pid):
+    cg_dir = get_cg_dir(dom_id)
+
+    try:
+        os.mkdir(cg_dir, 0755)
+    except OSError as e:
+        if e.errno != errno.EEXIST:
+            log.debug("Failed to create cgroup: {}".format(cg_dir))
+            exit(1)
+
+    try:
+        with open(cg_dir + "tasks", "w", 0) as tasks, \
+                open(cg_dir + "devices.deny", "w", 0) as deny, \
+                open(cg_dir + "devices.allow", "w", 0) as allow:
+
+            # deny all by default
+            deny.write("a")
 
-    if len(sys.argv) != 2:
-        sys.exit("usage: {} device_node".format(sys.argv[0]))
+            # Only the "/dev/null" is required after CP-25630
+            whitelist = [
+                ("/dev/null", "rw")
+            ]
+            for dev in whitelist:
+                ctl = get_ctl(*dev)
+                if ctl:
+                    allow.write(ctl)
 
-    device = sys.argv[1]
-    pattern = re.compile(r"^\d+-\d+(\.\d+)*$")
-    if pattern.match(device) is None:
-        log.debug("unexpected device node: {}".format(device))
+            tasks.write(str(pid))
+
+    except IOError as e:
+        log.debug("Failed to setup cgroup: {}".format(str(e)))
         exit(1)
 
+
+def mount(source, target, fs, flags=0):
+    if ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True
+                   ).mount(source, target, fs, flags) < 0:
+        # log and continue
+        log.debug("Failed to mount {} ({}) to {} with flags {}: {}".
+                  format(source, fs, target, flags,
+                         os.strerror(ctypes.get_errno())))
+
+
+def umount(target):
+    if ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True
+                   ).umount(target) < 0:
+        # log and continue
+        log.debug("Failed to umount {}: {}".
+                  format(target, os.strerror(ctypes.get_errno())))
+
+
+def attach(path, dom_id, pid, reset_only):
+    # reset device
     try:
-        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
-        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
-        reset_device(bus, dev)
-    except (IOError, ValueError):
-        log.debug(traceback.format_exc())
+        with open(path, "w") as f:
+            # USBDEVFS_RESET _IO('U', 20)
+            USBDEVFS_RESET = (ord('U') << 8) | 20
+            fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+    except IOError as e:
+        # log and continue
+        log.debug("Failed to reset {}: {}".format(path, str(e)))
+
+    if reset_only:
+        return
+
+    # set device file uid/gid
+    os.chown(path, pwd.getpwnam("qemu_base").pw_uid + dom_id,
```

Comment:
I think it would be better if the uid & gid should come from the caller. This will allow us to change it more easily in the future (which I think we should do since it currently sometimes overlaps with other users).

Reply:
TODO, requires toolstack change.

Reply:
I think this should be covered in a separate PR such that we can merge this.

Reply:
Sure, that's fine.